### ISSUE TITLE
[question] require dependency with any option value 

### ISSUE THREAD

**Issue by boussaffawalid** [conan-io/conan#11091]
Hello,

I have a dependency `libA` that have an option `with_feature`, something like:
```
class DemoConan(ConanFile):
    name = "libA"
    options = {"with_feature": [True, False]}
    default_options = {"with_feature": True}
```

And I have another recipe that depends on `libA`, but it does not matter if it has `with_feature` set to true or false.
Is there a way in conan to depends on any options values for a specific dependency .

I've checked the "ANY" option value from https://docs.conan.io/en/latest/reference/conanfile/attributes.html#options which could be used somehow, but that will require altering the recipe of libA to something like:

```
class DemoConan(ConanFile):
    name = "libA"
    options = {"with_feature": "ANY"}
    default_options = {"with_feature": True}

   def package_id(self):
        if self.option.with_feature == "with_feature":
            del del self.info.options.with_feature
```

Is there a better way todo this at the consumer's level ?

**Comment by memsharded** [comment]
Hi @boussaffawalid 

The idea is that if a package doesn't care about their dependencies options, they wouldn't define their default values at all. 

The removal of ``options`` in the ``package_id()`` for the package ``libA`` itself, would mean that the binary is identical both when built with the feature and without the feature, which seems unlikely, so doesn't sound a good approach. That is only ok, when the binary is identical, and you want the options mainly for controlling "info", like flags or similar that will get back to the consumer.

If the problem is a problem of conflicts that appear because some packages will be defining options for the dependencies different to the default ones, then, there is not much to do there, besides resolving the conflicts in the downstream consumers, specifying the final options themselves.

Please let me know if this help, I am not fully sure I understood the question.

**Comment by boussaffawalid** [comment]
@memsharded  well, the problem is that I have two consumers of `libA`. one require a specific value of  `with_feature` and the other any value of it. so removing the option from the `package_id` wont work

**Comment by memsharded** [comment]
> @memsharded well, the problem is that I have two consumers of libA. one require a specific value of with_feature and the other any value of it. so removing the option from the package_id wont work

Yes, that is what I was saying, removing the ``option`` in ``libA`` ``package_id()`` method doesn't seem like the correct approach, as the ``libA`` package binary is different depending on the option.

So, still not clear what is the issue is, it is a ``package_id`` issue of the consumer? You want to make the consumer ``package_id`` independent of the dependency ``option`` value? You can do something like ``self.info.options.pop("libA")`` in the ``package_id()`` method of the consumer, not in ``libA`` one.

**Comment by boussaffawalid** [comment]
We have have thousands of builds for `libA`, some have `with_feature` set to true, other have `with_feature` set to false.

Now, one specific consumer of `libA` want to a use a specific version of `libA` without caring about the value of with_feature.  any value of `with_feature` should fulfill the requirement without the need to rebuild libA. 

For example:
let's say we have two builds of libA
* libA/1.0 (only has with_feature=true)
* libA/2.0 (only has with_feature=fase)

* If the consumer sets `default_options = {"libA:with_feature": True}` it can only consume libA/1.0 but not `libA/2.0`
* If the consumer sets `default_options = {"libA:with_feature": false}` it can only consume libA/2.0 but not `libA/1.0`

I need something like:
`default_options = {"libA:with_feature": ANY}`

**Comment by memsharded** [comment]
> libA/1.0 (only has with_feature=true)

Do you mean that you only build the binary with ``with_feature=True``. Then, this is not an option of ``libA/1.0``, if it always build with the feature enabled, it is not an option, not something that can be changed? I still don't fully get it.

>If the consumer sets default_options = {"libA:with_feature": True} it can only consume libA/1.0 but not libA/2.0

If the consumer doesn't need any specific value for ``libA:with_feature``, why defining a default in the first case? Just don't define it, it is not necessary to always define a value for dependencies options (actually the usual is not defining it).
It is also not clear the ``it can only consume``. What does it mean, is it getting a "version conflict" because of this, or is it getting a "binary not found" because the binary doesn't exist for that configuration?

**Comment by boussaffawalid** [comment]
> Do you mean that you only build the binary with with_feature=True. Then, this is not an option of libA/1.0, if it always build with the feature enabled, it is not an option, not something that can be changed? I still don't fully get it.

We build both for both `with_feature=True` and `with_feature=False`, but not for all versions. Some versions are libA are only built with `with_feature=True`, some are only built `with_feature=False` , others are built for both

`with_feature` must be an option as the final package is not the same.

And we do have two consumers of libA:
*  Consumer 1: cares about the value `with_feature` and can only consume versions which where created with `with_feature=True`
*  Consumer 2: don't care about the value `with_feature` and can consume any version of libA.
 

> It is also not clear the it can only consume. What does it mean, is it getting a "version conflict" because of this, or is it getting a "binary not found" because the binary doesn't exist for that configuration?

I mean "binary not found" because the binary doesn't exist for that configuration

> If the consumer doesn't need any specific value for libA:with_feature, why defining a default in the first case? Just don't define it, it is not necessary to always define a value for dependencies options (actually the usual is not defining it).

If I dont define the value ` libA:with_feature` in the consumer, which one will be used ? True or False ? I assume it will use the default value defined in LibA, True in my case (`default_options = {"with_feature": True}`).  But still in this case, ` Consumer 2` wont be able to use versions created for only `with_feature` set to False.

**Comment by memsharded** [comment]
> I mean "binary not found" because the binary doesn't exist for that configuration

Ok, I think I start to get it. It seems the problem is that the option in ``libA`` is affecting the ``package_id`` of ``consumer2``. This can be avoided, in the ``package_id()`` of ``consumer2``, try to do something like:

```python
class Consumer2(ConanFile):
     name = "consumer2"
     ...
     def package_id(self):
          self.info.options.pop("libA")
```
That will make ``consumer2`` binary independent of the options defined in ``libA``. 

> If I dont define the value libA:with_feature in the consumer, which one will be used ? True or False ? I assume it will use the default value defined in LibA, True in my case (default_options = {"with_feature": True}). 

The one that is defined, either by the ``libA`` default or by other possible packages in the graph will be used. As ``consumer2`` doesn't care, it doesn't need to define any specific value, right? It will accept any possible value that ``libA`` has.

> But still in this case, Consumer 2 wont be able to use versions created for only with_feature set to False.

Yes, if you add the above ``package_id()`` to ``consumer2``.


A different story, but that doesn't seem your current issue, is when "configuration" conflicts are raised. Those happens when 2 different branches in a diamond graph define different values for the option. In that case it is the final consumer in the root of the diamond the one that should resolve the conflict specifying which is the right value, with ``libA:with_feature=xxxx``.

**Comment by boussaffawalid** [comment]
@memsharded I tried the approach above, but still does not work. I got this error:

```
ERROR: Missing binary: libA ...

libA... : WARN: Can't find a 'libA ....' package for the specified settings, options and dependencies:
- Settings: .....
- Options: with_feature=True...
```

My understanding is that package_id is used to compute the package id of consumer2 but it is not used in the dependency resolution! or did I miss something ?

**Comment by memsharded** [comment]
Hi @boussaffawalid 

Sorry that this was not followed up back then.

Now this ticket looks outdated, as Conan 2 changed many things regarding ``package_id`` computation.
I am closing this ticket as outdated, but please don't hesitate to create new tickets for any further question you might have. Thanks for your feedback!
